add support for mutated params

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #106146
* __->__ #106098
* #106092
* #106097

Previously, this didn't work because of the warmup run. Now that we do not run warmup, and then execution on one inductor invocation this works. llama inference 1.6->4.4

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

eellison committed:
add support for mutated params

[ghstack-poisoned]

eellison committed:
Update on "add support for mutated params"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

ezyang commented on torch/_inductor/scheduler.py:
```diff
@@ -923,6 +923,11 @@ def add_user(used_by_name, user_node, can_inplace=False):
 
+        inp_names = list(V.graph.graph_inputs.keys())
+        V.graph.mutated_input_idxs = [
+            inp_names.index(name) for name in V.graph.mutated_inputs
```
🚨 quadratic police

ezyang commented on torch/_inductor/compile_fx.py:
```diff
@@ -325,9 +325,17 @@ def compile_fx_inner(
+        # doesnt work for non-trees because the warmup run would apply mutation twice
+        if config.triton.cudagraph_trees:
+            has_mutation = not all(
+                idx < num_fixed for idx in compiled_graph.mutated_input_idxs
```
a little comment saying that this is for mutated params would help

eellison committed:
Update on "add support for mutated params"


Previously, this didn't work because of the warmup run. Now that we do not run warmup, and then execution on one inductor invocation this works. llama inference 1.6->4.4

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison committed:
Update on "add support for mutated params"


Previously, this didn't work because of the warmup run. Now that we do not run warmup, and then execution on one inductor invocation this works. llama inference 1.6->4.4

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

eellison commented:
@pytorchbot merge
